src/validation.cpp

```diff
@@ -166,18 +169,24 @@ class CChainState {
     bool DisconnectTip(CValidationState& state, const CChainParams& chainparams, DisconnectedBlockTransactions *disconnectpool);
 
     // Manual block validity manipulation:
-    bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIndex *pindex);
-    bool InvalidateBlock(CValidationState& state, const CChainParams& chainparams, CBlockIndex *pindex);
-    bool ResetBlockFailureFlags(CBlockIndex *pindex);
+    bool PreciousBlock(CValidationState& state, const CChainParams& params, const CBlockIndex *pindex);
+    bool InvalidateBlock(CValidationState& state, const CChainParams& chainparams, const CBlockIndex *pindex);
+    bool ResetBlockFailureFlags(const CBlockIndex *pindex);
 
     bool ReplayBlocks(const CChainParams& params, CCoinsView* view);
     bool RewindBlockIndex(const CChainParams& params);
     bool LoadGenesisBlock(const CChainParams& chainparams);
-
-    void PruneBlockIndexCandidates();
+    bool LoadChainTip(const CChainParams& chainparams);
 
     void UnloadBlockIndex();
 
+    ~CChainState() {
+        BlockMap::iterator it1 = mapBlockIndex.begin();
```

Comment:
Commit "mapBlockIndex only in validation.cpp, and const outside CChainState"

Should `AssertLockHeld(cs_main)`?